config.go

```diff
@@ -183,8 +183,13 @@ func (c *Config) Client() (*alks.Client, error) {
 		}
 	}
 
-	// check if the user is using a assume-role IAM admin session or MI.
-	if isValidIAM(cident.Arn, client) != true {
+	// Validate STS for IAM active.
+	validateSTS, err := client.IsIamEnabled("")
+	if err != nil {
+		return nil, err
+	}
```

Comment:
I like that you removed the helper method, but we need to be careful about letting errors escape here I think.  If the STS credentials are for a non-MI role, I think this will return an auth error, which might be difficult to interpret.

There are a couple scenarios that could play out here:

- Valid STS credentials for a non-MI role -> IAM credential error
- Valid STS credetnials for a MI role -> Good
- Valid STS credentials for an ALKS role (with IAM) -> Good
- Valid STS credentials for an ALKS role (without IAM) -> IAM credential error
- Expired STS credentials -> ???
- Other ALKS Failure -> ???

I think for simplicity we should just return the IAM error if we fail to make the API call, but I can see an argument for trying to be more picky.

Reply:
If the STS credentials are for a non-MI role and they're not IAM enabled, the endpoint will return a `false` value, and it should return the following error:
```
errors.New("Looks like you are not using ALKS IAM credentials. This will result in errors when creating roles. \n " +
			"Note: If using ALKS CLI to get credentials, be sure to use the '-i' flag. \n Please see https://coxautoinc.sharepoint.com/sites/service-internal-tools-team/SitePages/ALKS-Terraform-Provider---Troubleshooting.aspx for more information.")
```

Which is expected. As far as the other scenarios described such as expired STS, or another ALKS failure, this would need to get picked up at the following call: `alks.NewSTSClient(...)` since we generate new STS credentials before calling to validate, every time right? We're using the client that is returned always, so if something goes wrong, it is handled beforehand.

So I guess the TL;DR is - if anything isn't IAM enabled, we're handling this as a 200 `false` response, otherwise if the endpoint fails altogether, then this is a failure to authenticate to ALKS which I think should be caught by the `newSTSClient(...)` call or beforehand where we validate with `GetCallerIdentity(...)`

Reply:
Fair point about expired credentials.  Timing would need to be _very_ unlucky to pass the `GetCallerIdentity` API check and then fail on the first ALKS API call.  However, I do think it's possible that we'd fail this check with a non-MI role type.  The `isIAMEnabled` endpoint expects an authenticated request, but flow would fail for non-MI roles.  Perhaps the solution here then is to make the code just return the error if we get an API failure as well.  Something along this line perhaps:
```go
	// Validate STS for IAM active.
	validateSTS, err := client.IsIamEnabled("")
	if err != nil || validateSTS.IamEnabled != true {
		return nil, errors.New("Looks like you are not using ALKS IAM credentials. This will result in errors when creating roles. \n " +
			"Note: If using ALKS CLI to get credentials, be sure to use the '-i' flag. \n Please see https://coxautoinc.sharepoint.com/sites/service-internal-tools-team/SitePages/ALKS-Terraform-Provider---Troubleshooting.aspx for more information.")
	}
```

I am not 100% in love with how that flows, but I really hate duplicating code.  If you have an alternative implementation I am totally on board.

Reply:
Okay yeah now I am following what you're saying. I cannot think of an "elegant" solution to handle the error when it fails from a non-MI STS, other than what you posted. However... I want to keep an eye on Splunk for when these errors do occur and see if the error thrown can give me a hint at how we could handle this without affecting other possible error outcomes. Good catch & suggestion!